# --- T2-COPYRIGHT-NOTE-BEGIN ---
# This copyright note is auto-generated by ./scripts/Create-CopyPatch.
# 
# T2 SDE: package/.../silo/util.patch.cross
# Copyright (C) 2009 The T2 SDE Project
# 
# More information can be found in the files COPYING and README.
# 
# This patch file is dual-licensed. It is available under the license the
# patched project is licensed under, as long as it is an OpenSource license
# as defined at http://www.opensource.org/ (e.g. BSD, X11) or under the terms
# of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
# --- T2-COPYRIGHT-NOTE-END ---

--- silo/second/util.c.vanilla	2009-10-22 20:30:36.000000000 +0200
+++ silo/second/util.c	2009-10-22 20:43:12.000000000 +0200
@@ -1,6 +1,7 @@
 /* Tool to make gzipped second stage binary
    
    Copyright (C) 1996,1998 Jakub Jelinek
+   Copyright (C) 2009 Rene Rebe (rene@exactcode.de)
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
@@ -109,6 +110,16 @@
     exit(1);
 }
 
+void save_diffs (short* diffs, int n, FILE* s)
+{
+	int i;
+	for (i = 0; i < n; i++)
+	{
+		unsigned char buf[2] = { ((unsigned short)diffs[i] >> 8) & 0xff, (unsigned short)diffs[i]  & 0xff};
+		fwrite(buf, 2, 1, s);
+	}
+}
+
 int main(int argc, char **argv)
 {
     FILE *g, *h;
@@ -175,10 +186,10 @@
     save (g, second_start - rodata_end, 0);
     save (h, second_end - second_start, 2);
     save (g, end - second_end, 0);
-    fwrite(diffs[0], 2, ndiffs[0]+1, g);
-    fwrite(diffs[1], 2, ndiffs[1]+1, g);
-    fwrite(diffs[2], 2, ndiffs[2]+1, h);
-    fwrite(diffs[3], 2, ndiffs[3]+1, h);
+    save_diffs (diffs[0], ndiffs[0]+1, g);
+    save_diffs (diffs[1], ndiffs[1]+1, g);
+    save_diffs (diffs[2], ndiffs[2]+1, h);
+    save_diffs (diffs[3], ndiffs[3]+1, h);
     fclose (f);
     fclose (e);
     fclose (g);
